docs/source-fabric/advanced/compile.rst

```diff
@@ -0,0 +1,180 @@
+#################################
+Speed up models by compiling them
+#################################
+
+Compiling your PyTorch model can result in significant speedups, especially on the latest hardware such as NVIDIA GPUs.
+This guide shows you how to apply ``torch.compile`` correctly in your code.
+
+.. note::
+
+    This requires PyTorch >= 2.0.
+
+
+----
+
+
+***********************************
+Apply `torch.compile` to your model
+***********************************
+
+Compiling a model in a script together with Fabric is as simple as adding one line of code, calling :func:`torch.compile`:
+
+.. code-block:: python
+
+    import torch
+    import lightning as L
+
+    # Set up Fabric
+    fabric = L.Fabric(devices=1)
+
+    # Define the model
+    model = ...
+
+    # Compile the model
+    model = torch.compile(model)
+
+    # `fabric.setup()` should come after `torch.compile()`
+    model = fabric.setup(model)
+
+
+.. important::
+
+    You should compile the model **before** calling ``fabric.setup()`` as shown above for an optimal integration with features in Fabric.
+
+The newly added call to ``torch.compile()`` by itself doesn't do much. It just wraps the model in a "compiled model".
+The actual optimization will start when calling ``forward()`` on the model for the first time:
+
+.. code-block:: python
+
+    # 1st execution compiles the model (slow)
+    output = model(input)
+
+    # All future executions will be fast
+    output = model(input)
+    output = model(input)
+    ...
+
+This is important to know when you measure the speed of a compiled model and compare it to a regular model.
+You should always *exclude* the first call to ``forward()`` from your measurements, since it includes the compilation time.
+
+.. collapse:: Full example with benchmark
+
+    Below is an example that measures the speedup you get when compiling a DenseNet from TorchVision.
+
+    .. code-block:: python
+
+        import statistics
+        import torch
+        import torchvision.models as models
+        import lightning as L
+
+
+        @torch.no_grad()
+        def benchmark(model, input, num_iters=10):
+            """Runs the model on the input several times and returns the median execution time."""
+            start = torch.cuda.Event(enable_timing=True)
+            end = torch.cuda.Event(enable_timing=True)
+            times = []
+            for _ in range(num_iters):
+                start.record()
+                model(input)
+                end.record()
+                torch.cuda.synchronize()
+                times.append(start.elapsed_time(end) / 1000)
+            return statistics.median(times)
+
+
+        fabric = L.Fabric(accelerator="cuda", devices=1)
+
+        model = models.densenet121()
+        input = torch.randn(16, 3, 128, 128, device=fabric.device)
+
+        # Compile!
+        compiled_model = torch.compile(model, mode="reduce-overhead")
```

Comment:
I'm looking for a model that sees a speedup when you compile it with defaults (no cuda graphs). The current one here would actually be slower 😐